File: api/src/main/java/com/tersesystems/echopraxia/Level.java
package com.tersesystems.echopraxia;

/**
 * An enumeration of the logging levels.
 *
 * <p>Note that ordinal values may not match with Logback/Log4J/SLF4J levels, and you should
 * exercise caution when looking at cardinal/numeric values because they're all different.
 */
public enum Level {

  // Order is significant, and compareTo uses ordinal values internally.
  TRACE, // 0
  DEBUG, // 1
  INFO,  // 2
  WARN,  // 3
  ERROR; // 4

  public boolean isGreater(Level r) {
    return compareTo(r) > 0;
  }

  public boolean isGreaterOrEqual(Level r) {
    return compareTo(r) >= 0;
  }

  public boolean isLess(Level r) {
    return compareTo(r) < 0;
  }

  public boolean isLessOrEqual(Level r) {
    return compareTo(r) <= 0;
  }

  public boolean isEqual(Level r) {
    return equals(r);
  }
}

File: scala/api/src/main/scala/com/tersesystems/echopraxia/sapi/Level.scala
package com.tersesystems.echopraxia.sapi

import com.tersesystems.echopraxia.{Level => JLevel}

object Level {
  case object TRACE extends Level(com.tersesystems.echopraxia.Level.TRACE)
  case object DEBUG extends Level(com.tersesystems.echopraxia.Level.DEBUG)
  case object INFO  extends Level(com.tersesystems.echopraxia.Level.INFO)
  case object WARN  extends Level(com.tersesystems.echopraxia.Level.WARN)
  case object ERROR extends Level(com.tersesystems.echopraxia.Level.ERROR)

  def asScala(level: JLevel): Level = level match {
    case JLevel.TRACE => TRACE
    case JLevel.DEBUG => DEBUG
    case JLevel.WARN  => WARN
    case JLevel.INFO  => INFO
    case JLevel.ERROR => ERROR
  }
}

sealed class Level(private val level: JLevel) {
  def isGreater(r: Level): Boolean = level.compareTo(r.level) > 0

  def isGreater(r: JLevel): Boolean = level.compareTo(r) > 0

  def isGreaterOrEqual(r: Level): Boolean = level.compareTo(r.level) >= 0

  def isGreaterOrEqual(r: JLevel): Boolean = level.compareTo(r) >= 0

  def isLess(r: Level): Boolean = level.compareTo(r.level) < 0

  def isLess(r: JLevel): Boolean = level.compareTo(r) < 0

  def isLessOrEqual(r: Level): Boolean = level.compareTo(r.level) <= 0

  def isLessOrEqual(r: JLevel): Boolean = level.compareTo(r) <= 0

  def isEqual(r: Level): Boolean = this.level == r.level

  def isEqual(r: JLevel): Boolean = this.level == r

  def asJava: JLevel = level
}
